Only enable duplication when layer editable

signedav opened the pull request:
MapLayerActions and user actions has the attribute isEnabledOnlyWhenEditable.
In the user action it's configurable per checkbox in GUI. The MapLayerActions DuplicateFeature and DuplicateFeatureDigitized have it always active. User action Duplicate Selected Feature per default.

In the identify result panel the context menu only loads actions that are not onlyEnabledWhenEditable. Means DuplicateFeature is removed there.
In the attribute form dialog the action menu only loads actions that are not onlyEnabledWhenEditable or in editable mode and are not in AddFeatureMode. Means Duplicate Feature is removed on adding a feature.

Otherwise its disabled when the layer is not in editable mode. It cannot be selected because it's onlyEnabledWhenEditable.

In case there would be somewhere the possiblity to select it, althoug the layer is not editable, DuplicateFeature would not toggle editable mode and no duplication could be done.

Where you can see Duplicate Feature and Duplicate Feature Digitized MapLayer Actions disabled when not in editable mode:
- MapCanvas
- Attribute form actions of existing feature
- Attribute form relation widget rightclick
- Attribute list in formview rightclick 
- Attribute list in listview rightclick 
- Attribute list with actions in column
- Rightclick on feature on map 

Where you can see Duplicate Selected Features user action:
- Attribute list actions in toolbar

Where you can't see them at all:
- Identify result panel
- Attribute form actions of new feature

signedav committed:
disable or do not display maplayeractions

in some cases

signedav committed:
resolved conflicts

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -6452,8 +6452,12 @@ void QgisApp::refreshFeatureActions()
   QList<QgsAction> actions = vlayer->actions()->actions( QStringLiteral( "Canvas" ) );
   Q_FOREACH ( const QgsAction &action, actions )
   {
+    if( vlayer->readOnly() && action.isEnabledOnlyWhenEditable() )
```
I think this should be vlayer->isEditable(), not readOnly() - e.g. hide the actions if not editable as opposed to showing them as disabled. Best to keep the ui as clutter-free as possible when we can.

signedav replied:
My thought was: Hide them, when they never should appear. Disable them, when they would be available in editable mode. So the user sees, that there would be something, when he toggles edit mode (similar to the Tools like "Add Feature" etc.).
But I could change it to not even list them as well...

nyalldawson replied:
My personal preference would be to hide them. I really don't want to see actions become a huge mash of inbuilt features, since I think the primary use case for them is user-created actions, and I want to ensure that those user created actions remain prominent and aren't lost amongst many other inbuilt actions. 

@nirvn what do you think?

nirvn replied:
Yep agreed. I was about to write this very comment once at office.

andreasneumann replied:
@nyalldawson - sorry, but I disagree here. From the QGIS manual "QGIS provides the ability to perform an action based on the attributes of a feature. This can be used to perform * any number * of actions, for example, running a program with arguments built from the attributes of a feature or passing parameters to a web reporting tool." (https://docs.qgis.org/2.18/en/docs/user_manual/working_with_vector/vector_properties.html#actions-properties) - clearly this does not exclude internal QGIS functionality - if - it is based on a feature attribute. Which clearly is the case here. Based on a feature, we create a new feature, that has the same attributes. I don't think it is clearly defined that Actions are only for user-created actions, and not for builtin functionality. Plus - the Atlas features is also a builtin action that is always - just there.

The primary reason why Matthias and I suggested the "duplicate feature" functionality to be an action is the fact, that actions are - as far as I know - the only easy way to have right-click context menu functionality on table rows.
Plus - if something is defined as an action it appears at many places automatically (as buttons, in form menu, right-click on table rows, in action tool, identify tool, etc. without having to program a lot.

andreasneumann replied:
ok - I am probably wrong that actions are the only way to have a context menu. I see that other - non-action based entries are here.
But still it is convenient, that something defined as an Action appears at many places where one wants to do something based on a feature or based on a features attribute.

nyalldawson replied:
Yes, but you're coming from the viewpoint of someone who will use that feature heavily. For many others it's just extra interface noise cluttering the actions they've defined themselves. That's why I think we hide when they can't be used, because at least that helps mitigate the impact. 

I feel the same with the atlas now, but it was added when actions were only exposed in about half the places they are now. And the atlas functionality only shows in very narrow circumstances - you have to have a layout with an atlas defined on that layer. Thats in contrast to these two actions which are always present for every vector layer.

m-kuhn replied:
Interface cluttering vs. availability of functionality is always a tricky thing to get right (and probably impossible to completely get right due to the wide range of users QGIS has)

In this case here I feel, that hiding something based on layer editable state is only masquerading an issue and will not help. I personally often remember menu actions based on their position (e.g. 3rd from the bottom) and hiding and showing them based on the editable state breaks this workflow.

I am in favor of the approach proposed in here and if the concerns about too many actions continue, let's find a better solution like configuration options for enabling and disabling or nested menus.

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -6452,8 +6452,12 @@ void QgisApp::refreshFeatureActions()
+
     QString actionTitle = !action.shortTitle().isEmpty() ? action.shortTitle() : action.icon().isNull() ? action.name() : QStringLiteral( "" );
     QAction *qAction = new QAction( action.icon(), actionTitle, mFeatureActionMenu );
+    qAction->setEnabled( !action.isEnabledOnlyWhenEditable() || vlayer->isEditable() );
```
see note above

nyalldawson commented on src/gui/qgsmaplayeractionregistry.h:
```diff
@@ -50,13 +50,13 @@ class GUI_EXPORT QgsMapLayerAction : public QAction
      * \note using AllActions as a target probably does not make a lot of sense. This default action was settled for API compatibility reasons.
      */
-    QgsMapLayerAction( const QString &name, QObject *parent SIP_TRANSFERTHIS, Targets targets = AllActions, const QIcon &icon = QIcon() );
+    QgsMapLayerAction( const QString &name, QObject *parent SIP_TRANSFERTHIS, Targets targets = AllActions, const QIcon &icon = QIcon(), const bool enabledOnlyWhenEditable = false );
```
please use a enum flag instead of a bool here - it's much more future proof.

nyalldawson commented on src/gui/qgsmaplayeractionregistry.h:
```diff
@@ -77,6 +77,9 @@ class GUI_EXPORT QgsMapLayerAction : public QAction
     const Targets &targets() const {return mTargets;}
 
+    //! Return whether only enabled in editable mode
+    bool isEnabledOnlyWhenEditable() const { return mEnabledOnlyWhenEditable; }
```
Same here - use an  flags() enum and return that here instead of the bool getter

signedav committed:
Merge branch 'master' into uxduplication

nyalldawson commented on src/app/qgsfeatureaction.cpp:
```diff
@@ -82,8 +82,12 @@ QgsAttributeDialog *QgsFeatureAction::newDialog( bool cloneFeature )
       if ( !action.runable() )
         continue;
 
+      if( mLayer->readOnly() && action.isEnabledOnlyWhenEditable() )
```
Also here update to skip  the action if the layer is not isEditable(). (And the other places below too)

nyalldawson commented:
Nearly there, nice work!

> Rightclick on feature on map

Do you mean with the identify tool? If so, I think these actions should always be hidden there. The identify tool isn't a "data edit" tool, so the editing actions don't belong there.

signedav committed:
hide instead of disable

and use mode for identify - not to show the actions with isEnabledOnlyWhenEditable is active

signedav committed:
style stuff

signedav committed:
intagrated enumvalue IdentifyMode

used like SingleEditMode here

signedav committed:
document parameters

signedav committed:
skip enabledOnlyWhenEditable from pyhton bindings

in the construtors of QgsAction and QgsMapLayerAction

nyalldawson commented:
Ok, +1 for someone to merge if/when Travis is happy ( I'm off to bed!)

signedav commented:
Thanks @nyalldawson
